packages/kotti-ui/source/kotti-field-file-upload/KtFieldFileUpload.vue

```diff
@@ -0,0 +1,214 @@
+				@click.stop.prevent
+			>
+				<PreUploadedFileItem
+					v-for="fileInfo in preUploadedFilesList"
+					v-bind="{
+						...sharedProps,
+						fileInfo,
+					}"
+					:key="fileInfo.id"
+					@remove="onRemoveFile"
+				/>
+			</div>
+		</div>
+	</KtField>
+</template>
+
+<script lang="ts">
+import { computed, defineComponent, onBeforeMount } from '@vue/composition-api'
+
+import { KtField } from '../kotti-field'
+import { useField } from '../kotti-field/hooks'
+import { makeProps } from '../make-props'
+
+import DropArea from './components/DropArea.vue'
+import FileItem from './components/FileItem.vue'
+import PreUploadedFileItem from './components/PreUploadedFileItem.vue'
+import TakePhoto from './components/TakePhoto/TakePhoto.vue'
+import { KOTTI_FIELD_FILE_UPLOAD_SUPPORTS } from './constants'
+import { KottiFieldFileUpload, Shared } from './types'
+import { buildFileInfo, buildFileItem } from './utils/non-remote'
+import { validateFile } from './validators'
+
+export default defineComponent<KottiFieldFileUpload.PropsInternal>({
+	name: 'KtFieldFileUpload',
+	components: {
+		DropArea,
+		FileItem,
+		KtField,
+		PreUploadedFileItem,
+		TakePhoto,
+	},
+	props: makeProps(KottiFieldFileUpload.propsSchema),
+	setup(props, { emit }) {
+		const field = useField<KottiFieldFileUpload.ValueInternal>({
+			emit,
+			/**
+			 * We consider this field empty if:
+			 * a. No files are selected, or
+			 * b. All selected files are invalid
+			 */
+			isEmpty: (value) =>
+				value.length === 0 ||
+				value.every(
+					(fileItem) =>
+						validateFile({
+							extensions: props.extensions,
+							fileName: 'file' in fileItem ? fileItem.file.name : fileItem.name,
+							fileSize: 'file' in fileItem ? fileItem.file.size : fileItem.size,
+							maxFileSize: props.maxFileSize,
+						}) !== Shared.Validation.SUCCESS,
+				),
+			props,
+			supports: KOTTI_FIELD_FILE_UPLOAD_SUPPORTS,
+		})
+
+		const preUploadedFilesIds = new Set<KottiFieldFileUpload.FileInfo['id']>()
+
+		const filesList = computed<KottiFieldFileUpload.FileInfo[]>(() =>
+			field.currentValue
+				.filter((fileItem) => !preUploadedFilesIds.has(fileItem.id))
+				.map((fileItem) => {
+					const fileInfo = buildFileInfo({
+						extensions: props.extensions,
+						fileItem,
+						maxFileSize: props.maxFileSize,
+					})
+
+					if (fileItem.status !== fileInfo.status)
+						setStatus({ id: fileItem.id, status: fileInfo.status })
+
+					return fileInfo
+				}),
+		)
+		const showDropArea = computed(
+			() =>
+				props.allowMultiple ||
+				(!props.allowMultiple && field.currentValue.length === 0),
+		)
+
+		const setStatus = (payload: KottiFieldFileUpload.Events.SetStatus) => {
+			field.setValue(
+				field.currentValue.map((fileItem) =>
+					fileItem.id === payload.id
+						? { ...fileItem, status: payload.status }
+						: fileItem,
+				),
+				{ forceUpdate: true },
+			)
+		}
+
+		onBeforeMount(() => {
+			field.currentValue.forEach((fileItem) => {
+				if (
```

Comment:
shorter and more readable if
```
![undefined, Kotti.FieldFileUpload.Status.UPLOADED, Kotti.FieldFileUpload.Status.UPLOADED_WITH_ERROR].includes(fileItem.status)
```

up to you 

if you wrote it so ellaborately because of TS issues, then it makes sense, leave it as is :)

Reply:
right, i had that problem before when statuses were split and then merged